cloudmarker/clouds/gcpcloud.py

```diff
@@ -226,3 +192,112 @@ def done(self):
         up tasks associated with the :class:`GCPCloud` plugin. This
         may change in future.
         """
+
+
+def _get_normalized_firewall_rules(firewall):
+    """Split a firewall record into multiple firewall rules.
+
+    A firewall record in GCP has a top-level key named either
+    ``allowed`` or ``denied``. The value of this key is a list of
+    allowed or denied rules. Each such rule contains the name of allowed
+    or denied protocol along with a list of allowed or denied port
+    ranges.
+
+    In order to make it easier to write event plugins to detect security
+    issues in firewall, we generate a new firewall rule record for each
+    allowed or denied rule we find in the value of ``allowed`` or
+    ``denied`` keys in a firewall record.
+
+    Arguments:
+        firewall (dict): Raw firewall record retrieved from GCP.
+    """
+    allow_rules = firewall.get('allowed')
+    if allow_rules is not None:
+        for rule in allow_rules:
+            firewall_rule = _get_normalized_firewall_rule(firewall, rule)
+            firewall_rule['com']['access'] = 'allow'
+            yield firewall_rule
+
+    deny_rules = firewall.get('denied')
+    if deny_rules is not None:
+        for rule in deny_rules:
+            firewall_rule = _get_normalized_firewall_rule(firewall, rule)
+            firewall_rule['com']['access'] = 'allow'
+            yield firewall_rule
+
+
+def _get_normalized_firewall_rule(firewall, rule):
+    """Create a normalized firewall rule record.
+
+    Arguments:
+        firewall (dict): Raw firewall record retrieved from GCP.
+        rule (dict): Raw allowed or denied rule in ``firewall``.
+
+    Returns:
+        dict: A normalized firewall rule record with ``com`` bucket
+            populated with firewall rule properties in common notation.
+
+    """
+    record = {
+        'raw': rule,
+        'ext': {
+            'cloud_type': 'gcp',
+            'record_type': 'firewall_rule',
+            'firewall_id': firewall.get('id'),
+            'firewall_link': firewall.get('selfLink'),
+        },
+        'com': {
+            'cloud_type': 'gcp',
+            'record_type': 'firewall_rule',
+            'reference': firewall.get('selfLink'),
+
+            'enabled': not firewall.get('disabled'),
+
+            'direction':
+                _get_normalized_firewall_direction(firewall),
+
+            'protocol':
+                _get_normalized_firewall_protocol(rule),
+
+            'source_addresses': firewall.get('sourceRanges'),
+
+            # If the 'ports' key is missing in an allowed/denied rule,
+            # it means all ports are allowed/denied.
+            'destination_ports': rule.get('ports', ['0-65535'])
```

Comment:
How about using `*` in place of `0-65535`?

Reply:
Here's my rationale for choosing `0-65535` instead of `*`.

- In the standardized notation in `com` bucket, we want to simplify the notation, so if we can represent the notion of all ports with `0-65535`, there is no strict necessity of introducing a new notation like `*` to represent the same concept.
- A security analyst as well as event plugins would rely on the property names and values in `com` bucket to search and detect security issues. If we populate this `destination_ports` field with two different types of notation (an `M-N` notation as well as `*` notation), then the security analysts have to learn and use both notations while searching the `com` bucket and similarly our event plugin also needs to be able to support both types of notation. Choosing a single notation simplifies the work for both security analysts and event plugins.
- We already have two types of notation already in `destination_ports`: `M` notation (e.g, `80`) and `M-N` (e.g., `8000-8080`). These two notations are imposed on us by the existing data format in Azure on GCP. One can argue that we could have got rid of the `M` notation by writing `M` as `M-M` (i.e., writing `80` as `80-80` instead) but that is not a convenient or intuitive notation to work with for a human security analyst. So I decided to keep two separate notations: `M` notation for single port and `M-N` notation for port range. Since the notion of all ports is taken care of by `M-N` notation which is intuitive enough for the human user, I decided not to introduce a `*` notation.
- There is no specific benefit I see for supporting the `*` notation but there is at least one drawback (2nd point above).